File: week2_931/python-gdb-activities-20/activity12/gdb/domain/account_factory.py
# gdb/domain/account_factory.py
# ACTIVITY 12: Replace this file with your completed Activity 11 version before writing the tests.
from gdb.domain.iaccount import IAccount
from gdb.domain.savings_account import SavingsAccount
from gdb.domain.current_account import CurrentAccount
from gdb.domain.salary_account import SalaryAccount
from gdb.domain.fixed_deposit_account import FixedDepositAccount
from gdb.exceptions import AccountException

class AccountFactory:
    """Factory creating IAccount instances based on type."""
    @staticmethod
    def create_account(account_type: str, account_number: str, name: str, age: int, balance: float, status: str = "Active", pin: str = "0000") -> IAccount:
        # TODO (Step 2): Create and return the IAccount implementation matching account_type.
        #   - None/empty account_type -> raise AccountException
        #   - Normalise with account_type.strip().upper(), then map:
        #       "SAVINGS"      -> SavingsAccount(...)       with interest_rate=4.0, minimum_balance=1000.0
        #       "CURRENT"      -> CurrentAccount(...)       with overdraft_limit=10000.0
        #       "SALARY"       -> SalaryAccount(...)
        #       "FIXEDDEPOSIT" -> FixedDepositAccount(...)  with tenure_months=12, interest_rate=6.5
        #   - Any other value -> raise AccountException naming the unknown type
        raise NotImplementedError("TODO: implement AccountFactory.create_account()")

File: week2_931/java-gdb-activities-20/activity11/src/com/gdb/domain/AccountFactory.java
package com.gdb.domain;

public class AccountFactory {
    // TODO: Step 3 - Centralise account creation with a switch on the account type:
    //   - If type is null, return null.
    //   - switch (type.toUpperCase()):
    //       "SAVINGS"             -> a new SavingsAccount      (minBalance 1000.0, interestRate 4.0)
    //       "CURRENT"             -> a new CurrentAccount      (overdraftLimit 25000.0)
    //       "FIXED_DEPOSIT", "FD" -> a new FixedDepositAccount (tenureMonths 12, interestRate 6.5)
    //       "SALARY"              -> a new SalaryAccount       (employerName "TechCorp")
    //       default               -> throw new IllegalArgumentException("Unknown account type: " + type)
    public static IAccount createAccount(String type, String accNum, String name, int age, double balance, String status, String pin) {
        throw new UnsupportedOperationException("TODO: implement AccountFactory.createAccount (Activity 11, Step 3)");
    }
}
